src/copy/pvCopy.cpp

```diff
@@ -15,10 +15,11 @@
 #include <pv/pvData.h>
 #include <pv/bitSet.h>
 #include <pv/thread.h>
+#include "pv/pvPlugin.h"
 
 #define epicsExportSharedSymbols
-#include <pv/pvPlugin.h>
```

Comment:
Includes of headers being implemented must be below the epicsExportSharedSymbols.

```
../pvCopy.cpp:74:11: warning: ‘static epics::pvCopy::PVCopyPtr epics::pvCopy::PVCopy::create(const PVStructurePtr&, const PVStructurePtr&, const string&)’ redeclared without dllimport attribute: previous dllimport ignored [-Wattributes]
```

Is saying that that you're trying the provide an implementation for for something which has previously been declared as coming from another DLL.

Reply:
I did try this among all the other things I tried without success.
I am making another try.

Reply:
One other thing that often creates trouble:
Any .cpp source **must directly** `#include` all non-local EPICS include files for symbols that it uses, before `epicsExportSharedSymbols` is being defined. Indirect inclusion through other local header files screws up the DLL decoration.

Reply:
Where "local" means: inside the same DLL being created from the code in this .cpp file.

Reply:
FYI. the most reliable way I've found to find sharedLib.h import/export ordering issues is to look at the generated dependency .d file.  All the local headers (with relative paths) should be __after__ all of the non-local headers (with absolute paths).  This is stricter than necessary, but requires fewer CI iterations for me.

eg. in Base dbCa.c includes:

```c
#include "cadef.h"

#define epicsExportSharedSymbols
#include "db_access_routines.h"
```

https://github.com/epics-base/epics-base/blob/a64cb4a87775d0ea24cf46f18c1a6da7ce25bd83/modules/database/src/ioc/db/dbCa.c#L37-L43

In my build `modules/database/src/ioc/O.linux-x86_64/dbCa.d` contains:

```
dbCa.o: ../../../src/ioc/db/dbCa.c \
...
/home/mdavidsaver/work/epics/base-git/include/cadef.h \
 /home/mdavidsaver/work/epics/base-git/include/caerr.h \
 /home/mdavidsaver/work/epics/base-git/include/db_access.h \
 /home/mdavidsaver/work/epics/base-git/include/epicsTime.h \
 /home/mdavidsaver/work/epics/base-git/include/caeventmask.h \
 ../../../src/ioc/db/db_access_routines.h
```

So, I see only absolute paths included before `epicsExportSharedSymbols`, and only relative paths afterward (with the exception of compilerDependencies.h which doesn't import/export any symbols).

Reply:
There is one other thing I do (I think @mdavidsaver does the same but probably using a different mechanism) which is to set these build variables for Linux builds (but don't do this for Base-3.14):
```makefile
  OP_SYS_CFLAGS += -fvisibility=hidden
  OP_SYS_CXXFLAGS += -fvisibility-inlines-hidden
```
These compiler flags cause GCC to follow approximately the same rules for exporting symbols that Windows DLLs use, so I see the problems of incorrect exports on Linux builds and I know to correct them immediately. This also reduces the size of the generated lib.so files slightly since they don't have to export any symbols that aren't explicitly declared as being part of the library's public API.

I set the above and a few other things in my `~/configure` directory in a file named `CONFIG_USER.Common.linux-x86_64` which contains:
```makefile
# target-specific config for linux-x64 targets

# This is needed on RHEL
COMMANDLINE_LIBRARY = READLINE_NCURSES

OP_SYS_CFLAGS += -Wpointer-arith

ifdef BASE_3_15
  OP_SYS_CFLAGS += -fvisibility=hidden
  OP_SYS_CXXFLAGS += -fvisibility-inlines-hidden
endif
```
This ensures that all my Base builds will get those settings; don't set anything in there you don't want to be universal! There are other ways to do this too.

Reply:
I did this.
Thanks